server/src/controllers/ingredientsCtrlr.ts

```diff
@@ -0,0 +1,226 @@
+    return lower.slice(0, -1);
+  }
+
+  return lower;
+}
+
+// Make sure that the given string is a unit from the UNITS dictionary.
+function isUnit(word: string): boolean {
+  const lower = word.toLowerCase();
+
+  if (lower.endsWith('s')) {
+    return UNITS.has(lower.slice(0, -1));
+  }
+
+  return UNITS.has(lower);
+}
+
+// Remove any text between brackets in the given string.
+function removeTextInBrackets(input: string): string {
+  return input.replace(/\([^)]*\)/g, '').trim();
+}
+
+// Given a guaranteed single value that can be any of the following formats:
+//  - written fraction (ie: 1/2)
+//  - unicode vulgar fraction (ie: ½)
+//  - regular numbers (ie: 1 or 12 ...)
+// Return it in its corresponding type number. Return -1 on error.
+function convertSingleToNumber(single: string): number {
+  if (isNaN(Number(single))) {
+    if (single.includes('/')) {
+      const splitFrac = single.split('/');
+      const num = Number(splitFrac[0]);
+      const den = Number(splitFrac[1]);
+      if (splitFrac.length === 2 && !isNaN(num) && !isNaN(den)) {
+        return num / den;
+      }
+    } else if (single.length === 1) {
+      const norm = single.normalize('NFKD');
+      if (norm.length !== 1) {
+        const normValSplit = norm.split('\u2044');
+        if (normValSplit.length === 2) {
+          const num = parseFloat(normValSplit[0]);
+          const den = parseFloat(normValSplit[1]);
+          return num / den;
+        }
+      }
+    }
+    return -1;
+  } else {
+    return parseFloat(single);
+  }
+}
+
+// Convert the given string line into its corresponding type number.
+// The function takes care of unicode vulgar fraction instances and typed out fractions too.
+function convertLineToNumber(line: string): number {
+  const valSplit = line.split(' ');
+
+  if (valSplit.length === 1) {
+    const val = convertSingleToNumber(valSplit[0]);
+    if (val === -1) {
+      throw new Error(`Invalid number string sent: ${line}`);
+    }
+    return val;
+  } else if (valSplit.length === 2) {
+    const val1 = convertSingleToNumber(valSplit[0]);
+    const val2 = convertSingleToNumber(valSplit[1]);
+    if (val1 === -1 || val2 === -1) {
+      throw new Error(`Invalid number strings sent: ${line}`);
+    } else {
+      return val1 + val2;
+    }
+  } else {
+    throw new Error(`Unexpected value length: ${valSplit.length}`);
+  }
+}
+
+// Given an ingredient line from the list of ingredients in AllRecipe's web page, extract the data.
+const parseIngredientLine = (line: string): [string, number, string] => {
+  const lineSplit = line.split(' ');
+  if (lineSplit[1].length === 1) {
+    lineSplit[0] = `${lineSplit[0]} ${lineSplit[1]}`;
+    lineSplit.splice(1, 1);
+  }
+
+  let listItemIngredient = '';
+  let listItemQuantity = 0;
+  let listItemUnit = '';
+
+  if (lineSplit[1][0] === '(') {
+    try {
+      if (lineSplit[2].endsWith(')')) {
+        listItemQuantity = convertLineToNumber(lineSplit[1].replace('(', ''));

```

Comment:
Why do we need the `if` statement here? isn't this line also repeated in the `else` block

Reply:
It's not quite the same. In the else case I had to remove an item from the list and merge it with the previous one so that the flow may continue.